controllers/operatorscaler_controller.go

```diff
@@ -0,0 +1,334 @@
+			deploymentSpec := &csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[i].Spec
+			if *deploymentSpec.Replicas < int32(1) {
+				deploymentSpec.Replicas = ptr.To(int32(1))
+				updateRequired = true
+			}
+		}
+
+		if updateRequired {
+			if err := r.Client.Update(r.ctx, csv); err != nil {
+				r.log.Error(err, "failed updating csv replica", "csvName", csv.Name)
+				multierr.AppendInto(&returnErr, err)
+			} else {
+				r.log.Info("csv updated successfully", "csvName", csv.Name)
+			}
+		}
+	}
+
+	return returnErr
+}
+
+func (r *OperatorScalerReconciler) reconcileDynamicWatchers() error {
+
+	for i := range ResourceMappingList {
+		resourceMapping := ResourceMappingList[i]
+
+		if !r.kindsBeingWatched[resourceMapping.Kind] {
+
+			crd := &extv1.CustomResourceDefinition{}
+			if err := r.Client.Get(r.ctx, types.NamespacedName{Name: resourceMapping.CrdName}, crd); err != nil {
+				if !errors.IsNotFound(err) {
+					r.log.Error(err, "failed getting crd", "crdName", resourceMapping.CrdName)
+					return err
+				}
+			} else {
+				r.log.Info("adding dynamic watch", "kind", resourceMapping.Kind)
+
+				err := r.controller.Watch(
+					source.Kind(
+						r.cache,
+						client.Object(
+							&metav1.PartialObjectMetadata{
+								TypeMeta: metav1.TypeMeta{
+									APIVersion: resourceMapping.ApiVersion,
+									Kind:       resourceMapping.Kind,
+								},
+							},
+						),
+						&handler.EnqueueRequestForObject{},
+						// Trigger the reconcile for creation events of the object.
+						// This ensures the replicas in the CSV are scaled up based on the presence of Custom Resource (CR).
+						createOnlyPredicate,
+						predicate.GenerationChangedPredicate{},
+					),
+				)
+				if err != nil {
+					r.log.Error(err, "failed adding dynamic watch", "kind", resourceMapping.Kind)
+					return err
+				}
+
+				r.kindsBeingWatched[resourceMapping.Kind] = true
+			}
+		}
+	}
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OperatorScalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		Named("operatorScaler").
+		Watches(
+			&extv1.CustomResourceDefinition{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(
+				predicate.NewPredicateFuncs(func(obj client.Object) bool {
+					return slices.ContainsFunc(ResourceMappingList, func(resourceMapping ResourceMappingRecord) bool {
+						return resourceMapping.CrdName == obj.GetName()
+					})
+				}),
+				// Trigger a reconcile only during the creation of a specific CRD to ensure it runs exactly once for that CRD.
+				// This is required to dynamically add a watch for the corresponding Custom Resource (CR) based on the CRD name.
+				// The reconcile will be triggered with the CRD name as `req.Name`, and the reconciler will set up a watch for the CR using the CRD name.
+				createOnlyPredicate,
+				predicate.GenerationChangedPredicate{},
```

Comment:
This is irrelevant, there cannot be a generation change if you are only watching create events